src/vellum/workflows/nodes/displayable/tool_calling_node/utils.py

```diff
@@ -102,16 +102,21 @@ def create_port_condition(fn_name):
         # If no functions exist, create a simple Ports class with just a default port
         Ports = type("Ports", (), {"default": Port(default=True)})
 
-    # Add a chat history block to blocks
-    blocks.append(
-        VariablePromptBlock(
-            block_type="VARIABLE",
-            input_variable="chat_history",
-            state=None,
-            cache_config=None,
-        )
+    # Add a chat history block to blocks only if one doesn't already exist
+    has_chat_history_block = any(
+        hasattr(block, "input_variable") and block.input_variable == "chat_history" for block in blocks
```

Comment:
- use `block.block_type === "VARIABLE"` instead of hasattr
- pull out the hard coded string `"chat_history"` into a constant and replace all usages in this file that you find